lib/DotenvPlugin.js

```diff
@@ -257,16 +261,18 @@ class DotenvPlugin {
 			const { parsed, snapshot: newSnapshot } = dir
 				? await this._loadEnv(compiler, itemCache, dir)
 				: { parsed: {} };
-			// Merge parsed env variables with `process.env`
 			const env = this._getEnv(prefixes, parsed);
 
-			// update the definitions
-			definePlugin.definitions = envToDefinitions(env || {});
-			// update the file dependencies
+			envCache = env;
 			snapshot = newSnapshot;
 		});
 
-		compiler.hooks.compilation.tap(PLUGIN_NAME, (compilation) => {
+		compiler.hooks.thisCompilation.tap(PLUGIN_NAME, (compilation) => {
```

Comment:
We can't use `thisCompilation` here, because we need to replace it in child compilation too...